src/solver/expressions/include/antares/solver/expressions/nodes/TwoNodesVisitor.h

```diff
@@ -0,0 +1,98 @@
+/*
+** Copyright 2007-2024, RTE (https://www.rte-france.com)
+** See AUTHORS.txt
+** SPDX-License-Identifier: MPL-2.0
+** This file is part of Antares-Simulator,
+** Adequacy and Performance assessment for interconnected energy networks.
+**
+** Antares_Simulator is free software: you can redistribute it and/or modify
+** it under the terms of the Mozilla Public Licence 2.0 as published by
+** the Mozilla Foundation, either version 2 of the License, or
+** (at your option) any later version.
+**
+** Antares_Simulator is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** Mozilla Public Licence 2.0 for more details.
+**
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+#pragma once
+#include <optional>
+#include <vector>
+
+#include <antares/logs/logs.h>
+#include <antares/solver/expressions/nodes/NodesForwardDeclaration.h>
+
+namespace Antares::Solver::Nodes
+{
+namespace
+{
+template<class RetT, class VisitorT, class NodeT>
+std::optional<RetT> tryType(const Node& node1, const Node& node2, VisitorT& visitor)
+{
+    auto x = dynamic_cast<const NodeT*>(&node1);
+    auto y = dynamic_cast<const NodeT*>(&node2);
+    if (x && y)
+    {
+        return visitor.visit(*x, *y);
+    }
+    else
+    {
+        return {};
+    }
+}
+} // namespace
+
+template<class R>
+class TwoNodesVisitor
+{
+public:
+    virtual ~TwoNodesVisitor() = default;
+
+    R dispatch(const Node& node1, const Node& node2)
+    {
+        using FunctionT = std::optional<R> (*)(const Node&, const Node&, TwoNodesVisitor<R>&);
```

Comment:
What about different node types ? Is there a default case for this ? For example `AddNode` and `MultiplicationNode`.

Reply:
yes i have to check node consistency